cmd/libsnap-confine-private/bpf-support.c

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "bpf-support.h"
+
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+#include "utils.h"
+
+static int sys_bpf(enum bpf_cmd cmd, union bpf_attr *attr, size_t size) {
+#ifdef SYS_bpf
+    return syscall(SYS_bpf, cmd, attr, size);
+#else
+	errno = ENOSYS;
+    return -1;
+#endif
+}
+
+#define __ptr_as_u64(__x) ((uint64_t)(uintptr_t)__x)
+
+int bpf_create_map(enum bpf_map_type type, size_t key_size, size_t value_size, size_t max_entries) {
+    debug("create bpf map of type 0x%x, key size %zu, value size %zu, entries %zu", type, key_size, value_size,
+          max_entries);
+    union bpf_attr attr;
+    memset(&attr, 0, sizeof(attr));
+    attr.map_type = type;
+    attr.key_size = key_size;
+    attr.value_size = value_size;
+    attr.max_entries = max_entries;
+    return sys_bpf(BPF_MAP_CREATE, &attr, sizeof(attr));
+}
+
+int bpf_update_map(int map_fd, const void *key, const void *value) {
```

Comment:
In all other methods you added a `debug()` line at the beginning.

Reply:
Yeah, other operations are run mostly once, but update is called on each device allow which made it quite noisy when we set up all devices.